src/uct/ib/base/ib_device.h

```diff
@@ -151,8 +152,80 @@ typedef struct uct_ib_device_spec {
 } uct_ib_device_spec_t;
 
 
+/* Calculate real GIDs len. Can be either 16 (RoCEv1 or RoCEv2/IPv6)
+ * or 4 (RoCEv2/IPv4). This len is used for packets filtering by DGIDs.
+ *
+ * According to Annex17_RoCEv2 (A17.4.5.2):
+ * "The first 40 bytes of user posted UD Receive Buffers are reserved for the L3
+ * header of the incoming packet (as per the InfiniBand Spec Section 11.4.1.2).
+ * In RoCEv2, this area is filled up with the IP header. IPv6 header uses the
+ * entire 40 bytes. IPv4 headers use the 20 bytes in the second half of the
+ * reserved 40 bytes area (i.e. offset 20 from the beginning of the receive
+ * buffer). In this case, the content of the first 20 bytes is undefined." */
+static UCS_F_ALWAYS_INLINE
+unsigned uct_ib_gid_calc_len(const union ibv_gid *gid)
+{
+    /* Make sure that daddr in IPv4 resides in the last 4 bytes in GRH */
+    UCS_STATIC_ASSERT((UCT_IB_GRH_LEN - (20 + offsetof(struct iphdr, daddr))) ==
+                      sizeof(struct in_addr));
+
+    /* Make sure that dgid resides in the last 16 bytes in GRH */
+    UCS_STATIC_ASSERT((UCT_IB_GRH_LEN - offsetof(struct ibv_grh, dgid)) ==
+                      sizeof(struct in6_addr));
+
+    /* IPv4 mapped to IPv6 looks like: 0000:0000:0000:0000:0000:ffff:????:????,
+     * so check for leading zeroes and verify that 11-12 bytes are 0xff.
+     * Otherwise either RoCEv1 or RoCEv2/IPv6 are used. */
+    if ((gid->raw[0] == 0x00) && (gid->raw[1] == 0x00)) {
+        ucs_assert_always((gid->raw[10] == 0xff) && (gid->raw[11] == 0xff));
+        return sizeof(struct in_addr);
+    } else {
+        return sizeof(struct in6_addr);
+    }
+}
+
 KHASH_TYPE(uct_ib_ah, struct ibv_ah_attr, struct ibv_ah*);
 
+/* device GIDs set */
+KHASH_TYPE(uct_ib_gid, union ibv_gid, char);
+
+static UCS_F_ALWAYS_INLINE
+khint32_t uct_ib_kh_gid_hash_func(union ibv_gid gid)
+{
+    long len  = uct_ib_gid_calc_len(&gid);
+    void *ptr = UCS_PTR_BYTE_OFFSET((&gid) + 1, -len);
+    uint32_t gid32;
+    uint64_t gid64;
+
+    switch (len) {
+    case sizeof(uint32_t):
```

Comment:
imo, better use defined `ADDR_LEN` values and assert here that it is `sizeof(32/64)`